src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.cs

```diff
@@ -1478,6 +1456,84 @@ private static bool GetIsDaylightSavingsFromUtc(DateTime time, int year, TimeSpa
+
+            // DST is not supported if the year doesn't have a rule.
+            if (nextYearRule is null)
+            {
+                return default;
+            }
+
+            DaylightTimeStruct nextdaylightTime;
+            // Check if the rule not specifying the DST start as Jan 1st, 12:00 AM.
+            if (!nextYearRule.IsStartDateMarkerForBeginningOfYear())
+            {
+                nextdaylightTime = zone.GetDaylightTime(nextYear, nextYearRule, nextYearRuleIndex);
+
+                // Check if we are dealing with a southern sphere time zone having a DST start date after the end date.
+                // If the rule specifies the DST end as of Jan 1st, 12:00 AM that means the year will end with DST on
+                // but also means the year not started with DST as we already checked that before.
+                if (nextdaylightTime.Start < nextdaylightTime.End || nextYearRule.IsEndDateMarkerForEndOfYear())
+                {
+                    return default;
+                }
+
+                // It is the Southern sphere time zone. The year is started with DST on. Use the DST end to get the when DST ends that year.
+                return nextdaylightTime.End - utc - nextYearRule.BaseUtcOffsetDelta - nextYearRule.DaylightDelta;
+            }
+
+            // Comes here if not a southern sphere time zone.
+
+            // Check if DST ends specified as Jan 1st, 12:00 AM which means the year will end with DST on.
+            if (nextYearRule.IsEndDateMarkerForEndOfYear())
+            {
+                return new DateTime(nextYear, 12, 31) - utc - nextYearRule.BaseUtcOffsetDelta - nextYearRule.DaylightDelta;
+            }
+
+            // Not southern sphere time zone and, the year doesn't end with DST. Get the DST end date regularly as specified in the rule.
+            nextdaylightTime = zone.GetDaylightTime(nextYear, nextYearRule, nextYearRuleIndex);
+            return nextdaylightTime.End - utc - nextYearRule.BaseUtcOffsetDelta - nextYearRule.DaylightDelta;
+        }
+
+        // This method checks if a specific year end with DST, if so, will return when the DST starts that year.
+        // Otherwise will return a default DateTime value.
+        private static DateTime GetStartOfDstIfYearEndWithDst(int previousYear, TimeSpan utc, TimeZoneInfo zone)
+        {
+            AdjustmentRule? previousYearRule = zone.GetAdjustmentRuleForTime(new DateTime(previousYear, 12, 31), out int? previousYearRuleIndex);
+
+            // DST is not supported if the year doesn't have a rule.
+            if (previousYearRule is null)
+            {
+                return default;
+            }
+
+            DaylightTimeStruct previousDaylightTime;
+            // Check if the rule not specifying the DST end as Jan 1st, 12:00 AM.
+            if (!previousYearRule.IsEndDateMarkerForEndOfYear())
```

Comment:
Is this the equivalent logic?

```C#
            DaylightTimeStruct previousDaylightTime = zone.GetDaylightTime(previousYear, previousYearRule, previousYearRuleIndex);

            // If the rule is not specifying that the year ends with DST on and is in the Northern hemisphere, don't calculate the start of DST here
            if (!previousYearRule.IsEndDateMarkerForEndOfYear() && previousDaylightTime.Start < previousDaylightTime.End)
            {
                return default;
            }

            return previousDaylightTime.Start - utc - previousYearRule.BaseUtcOffsetDelta;
```

Reply:
Or maybe even inverting the condition would make it even more readable:

```C#
            DaylightTimeStruct previousDaylightTime = zone.GetDaylightTime(previousYear, previousYearRule, previousYearRuleIndex);

            // If the rule is specifying that the year ends with DST on or DST starts after it ends for the year (i.e. it is in the Southern hemisphere), calculate the time DST started
            if (previousYearRule.IsEndDateMarkerForEndOfYear() && previousDaylightTime.End < previousDaylightTime.Start)
            {
                return previousDaylightTime.Start - utc - previousYearRule.BaseUtcOffsetDelta;
            }

            return default;
```

Reply:
did you mean to use || instead of && in the condition?

it is possible to write it the way you suggest

Reply:
> did you mean to use || instead of && in the condition?

On the 2nd one - yes.  Oops - coding in PR comments is hard.  I've updated my snippet above.